### PR TITLE
Audio modulating support

### PR DESCRIPTION
**Pull request by microsoftsam**
Support for flashing using the ltc audio modulator (MIT).

### CODE REVIEW

File: docs/static/bubble.svg [hunk 1]
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<!-- Created with Inkscape (http://www.inkscape.org/) -->
+
+<svg
```
Comment (pelikhan) [inline docs/static/bubble.svg]: clean up svg?

File: webapp/src/app.tsx [hunk 2]
```diff
@@ -934,6 +934,13 @@ export class ProjectView
         if (/webusb=1/i.test(window.location.href)) {
             pxt.usb.initAsync().catch(e => { })
         }
+        // if the audio modulator is used, play silence first
+        if (pxt.appTarget.compile.useModulator) {
```
Comment (pelikhan) [inline webapp/src/app.tsx]: compile might be undefined

File: webapp/src/app.tsx [hunk 3]
```diff
@@ -1549,6 +1556,7 @@ ${compileService ? `<p>${lf("{0} version:", "C++ runtime")} <a href="${Util.html
         const trace = run && simOpts.enableTrace;
         const fullscreen = run && !inTutorial && !simOpts.hideFullscreen
         const audio = run && !inTutorial && targetTheme.hasAudio;
+        const useModulator = pxt.appTarget.compile.useModulator;
```
Comment (pelikhan) [inline webapp/src/app.tsx]: compile might be undefined

File: webapp/src/cmds.ts [hunk 4]
```diff
@@ -1,9 +1,13 @@
 /// <reference path="../../built/pxtlib.d.ts"/>
+/// <reference path="../../localtypings/modulator.d.ts"/>
 import * as core from "./core";
 import * as pkg from "./package";
 import * as hwdbg from "./hwdbg";
 import * as hidbridge from "./hidbridge";
 import Cloud = pxt.Cloud;
+import ModControllerConstructor = require("chibitronics-ltc-modulate");
```
Comment (pelikhan) [inline webapp/src/cmds.ts]: require lazily

File: webapp/src/cmds.ts [hunk 5]
```diff
@@ -31,6 +35,125 @@ export function browserDownloadDeployCoreAsync(resp: pxtc.CompileResult): Promis
             resp.userContextWindow,
             e => core.errorNotification(lf("saving file failed..."))
         );
+    } else if (pxt.appTarget.compile.useModulator) {
```
Comment (pelikhan) [inline webapp/src/cmds.ts]: can compile be undefined?

File: webapp/src/cmds.ts [hunk 6]
```diff
@@ -31,6 +35,125 @@ export function browserDownloadDeployCoreAsync(resp: pxtc.CompileResult): Promis
...
+
+        let bin = ltcIhexToBinary(resp.outfiles[pxtc.BINARY_HEX]);
+
+        let hex = resp.outfiles[pxtc.BINARY_HEX];
+
+        function getCanvas(): HTMLCanvasElement {
+            return document.getElementById('wavStrip') as HTMLCanvasElement;
+        }
+        function getAudioElement(): HTMLAudioElement {
+            return document.getElementById('audio_output') as HTMLAudioElement;
```
Comment (pelikhan) [inline webapp/src/cmds.ts]: could we get a consistent naming for the modulator HTML DOM element, modulelatorAudioOutput...

File: webapp/src/cmds.ts [hunk 7]
```diff
@@ -31,6 +35,125 @@ export function browserDownloadDeployCoreAsync(resp: pxtc.CompileResult): Promis
...
+                strip.stroke();
+                lastSample = sample;
+            }
+        }
+
+        if (modController) {
+            modController.stop();
+        }
+
+        modController = new ModControllerConstructor({
```
Comment (pelikhan) [inline webapp/src/cmds.ts]: require LTC modulator here

File: webapp/src/cmds.ts [hunk 8]
```diff
@@ -142,7 +299,7 @@ export function initCommandsAsync(): Promise<void> {
     } else if (pxt.winrt.isWinRT()) { // window app
         pxt.commands.deployCoreAsync = pxt.winrt.deployCoreAsync;
         pxt.commands.browserDownloadAsync = pxt.winrt.browserDownloadAsync;
-    } else if (Cloud.isLocalHost() && Cloud.localToken && !forceHexDownload) { // local node.js
+    } else if (Cloud.isLocalHost() && Cloud.localToken && !forceHexDownload && !pxt.appTarget.compile.useModulator) { // local node.js
```
Comment (pelikhan) [inline webapp/src/cmds.ts]: can compile be undefined?

**Comment by microsoftsam** [comment]
Closing to move ltc logic to target

**Pull request closed without merging**